sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -80,6 +80,90 @@ trait ScalaReflection {
     case (other, _) => other
   }
 
+  /**
+   * Creates a converter function that will convert Scala objects to the specified catalyst type.
+   */
+  private[sql] def createCatalystConverter(dataType: DataType): Any => Any = {
+    def extractOption(item: Any): Any = item match {
+      case o: Some[_] => o.get
+      case other => other
+    }
+
+    dataType match {
+      // Check UDT first since UDTs can override other types
+      case udt: UserDefinedType[_] =>
+        (item) => {
+          if (item == None) null else udt.serialize(extractOption(item))
```

Comment:
I won't be able to do something like `extractOption(item).asInstanceOf[Map[_, _]]` (or `extractOption(item).f()`) in general. Unless I compare twice, of course.

True, handling null separately for UDT's might be a good idea. I'll do that.